### ISSUE TITLE
StackOverflowException when circular dependency

### ISSUE THREAD

**Issue by jinek** [unitycontainer/unity#225]
Under some circumstances (but when circular reference exists) unity goes in to stackoverflow:

![image](https://user-images.githubusercontent.com/10516222/40054561-8cc242fa-5845-11e8-9ac2-c994cf52cca4.png)

Stacktrace does not contain user code.

**Comment by ENikS** [comment]
It is kind of logical behavior, isn't it? 
For me to fix it you might want to provide more information. Perhaps even a test case I could run?

**Comment by jinek** [comment]
I think it is almost necessary to notify caller by appropriate exception with an message (the only way to say caller about situation). StackOverflowException does not go to catch that mean it cant be logged.

A little bit later I will try to simplify solution so I will be able to share it.

**Comment by jinek** [comment]
May be it should complete each other? Test will give same information as a runtime, right?

At this moment unity produces exception that is not covered by documentation, that is typically (for .NET application) understood as an issue in this component. If you consider StackOverflowException as expected result it is also expected that its documented. But better is to avoid this exception and throw "CircularReferenceException" or something, because it will allow user to have more information about the state in which this situation occurred.

**Comment by ENikS** [comment]
Are you planning to provide unit test so I could fix it? If not, I will be closing this issue.

**Comment by jinek** [comment]
Sorry for delay.
I reproduced it in console application, hope it will be fine:

` internal class Program
    {
        public static void Main(string[] args)
        {
            var unityContainer = new UnityContainer();
            unityContainer.RegisterType<I1, C1>();
            unityContainer.RegisterType<I2, C2>();

            //next line returns StackOverflowException
            unityContainer.Resolve<I2>();
        }
    }

    public interface I1{}
    public interface I2{}

    public class C1 : I1{public C1(I2 i2){}}

    public class C2 : I2{public C2(I1 i1){}}`

**Comment by jinek** [comment]
![image](https://user-images.githubusercontent.com/10516222/40492581-138a58ea-5f71-11e8-86c3-669ec63f5bb9.png)

**Comment by jinek** [comment]
May be let me fix it by introducing stack depth counter?

**Comment by s-bauer** [comment]
@ENikS wouldn't it be possible to analyse the dependency graph before building it up and checking for circular dependencies? It should not be that difficult I think. I need to get familiar with the container code, then I might be able to give it a try myself.

**Comment by ENikS** [comment]
It is rather time consuming in current implementation and should not be done in production. 
This one should be handled by developers in testing stage.

**Comment by s-bauer** [comment]
Hmm I get that. How about a `[Conditional("DEBUG")]` function then?

**Comment by ENikS** [comment]
Yes, something alone these lines. But remember, nuget packages are all release, so DEBUG condition will not do mach for average user. Got to be something configurable...

**Comment by s-bauer** [comment]
I think the [ConditionalAttribute](https://msdn.microsoft.com/de-de/library/system.diagnostics.conditionalattribute(v=vs.110).aspx) should work. It checks for these flags on runtime.

**Comment by ENikS** [comment]
Not a bad idea, I'll think what could be done.

**Comment by s-bauer** [comment]
I've got it wrong. The ConditionalAttribute is also compile time :/

**Comment by ENikS** [comment]
I could create a public method on UnityContainer itself:
```C#
[Conditional("DEBUG")]
public UnityContainer EnableValidation();
```
the usage will be like this:
```C#
_container = new UnityContainer()
    .EnableValidation();
    .RegisterType(...)
```
it will execute only if client code executed in DEBUG context and will be skipped in all other cases.

I'll try if it works...

**Comment by s-bauer** [comment]
Basically something along these lines should work. However `Conditional` functions cannot have a return type, so you cannot integrate it that nicely in the fluent interface.

**Comment by jinek** [comment]
Idea that tests will be always executed in DEBUG compilation won't work. However is it possible to create StackFrame and check which attributes were applied to methods in stack, but it will be performance issue.. Why don't you want to just replace StackOverflowException with any another which can be caught?

**Comment by ENikS** [comment]
The problem with this exception: it is not thrown by Unity. It is OS that catches and throws it.

> Idea that tests will be always executed in DEBUG compilation won't work.

 Can you elaborate on this?

**Comment by ENikS** [comment]
[Moved](https://github.com/unitycontainer/container/issues/105)

**Comment by jinek** [comment]
> Can you elaborate on this?


Some developers execute tests in Release mode, because of wish to test exactly what will go to production pipeline/

**Comment by ENikS** [comment]
I think it is a solution in search for a problem. If testing is done right, there is no need for validation in production.

I am thinking of creating two versions of ``EnableValidation`` method. 
- IUnityContainerConfiguration.EnableValidation() - will enable validation regardless of environment
- UnityContainer.EnableDebugValidation() - will only validate in DEBUG build

**Comment by ENikS** [comment]
> Same can be archived by checking graph before or during build, that does not matter much, my idea is we should not make this exception configurable, it should be always thrown independently of compilation or environment options. Otherwise, as you said, OS will throw StackOverflowException (which can not be caught and logged).

You can't reliably analyze graph in expression building phase. In some cases it is possible to have reference to itself (within the graph) if PerResolveLifetimeManager is used for that type. It can only be determined at runtime.

**Comment by jinek** [comment]
Yeah, I thought same. Stack depth check will work.

**Comment by ENikS** [comment]
For Mono apps it might not work either. They have rather shallow stacks, how are you going to work around that?

**Comment by jinek** [comment]
Just put lower limit on dependency depth in Unity.
As I said in that comment, **MaxDependencyDepth** can be made configurable. For mono it can can be 50, for Xamarin 30, for custom CLR host can be customized by developer. pm me somewhere, we can discuss it and put result here, for example: jinek@msn.com

**Comment by ENikS** [comment]
I'll reopen the issue so others can participate. I think what you proposing could work but I would love to find more elegant solution.

**Comment by ENikS** [comment]
In reality it is not that difficult to walk back the chin, just use [IBuilderContext.ParentContext](https://github.com/unitycontainer/abstractions/blob/master/src/Builder/IBuilderContext.cs). But to do it properly you need to look at LifettimeManager as well, if it is either PerResolve or already initialized with value. So, as I said before, time consuming...
